### PR TITLE
feat: ros2 connector

### PR DESCRIPTION
**Pull request by maciejmajek**
## Purpose

New RAI sdk is based on Connector architecture which closely resembles ROS 2 features/api. This PR implements framework standard ROS2 connectors. 

## Proposed Changes

ROS2 ARI (Agent-Robot-Interface) connector based on out generic ROS 2 api. 

## Issues

- Links to relevant issues

## Testing

- How was it tested, what were the results?

## Summary by CodeRabbit

- **New Features**
	- Enhanced ROS2 communication framework with improved message handling.
	- Added flexible message initialization with metadata support.
	- Introduced new ROS2 connector for advanced messaging capabilities.

- **Bug Fixes**
	- Improved error handling in message and action processing.
	- Refined topic and message type verification mechanisms.

- **Refactor**
	- Simplified message class structure by removing unnecessary classes.
	- Centralized test helper utilities for ROS2 communication.
	- Updated message and action handling methods.

- **Tests**
	- Added comprehensive test suite for ROS2 communication components.
	- Implemented test cases for message sending, receiving, and action management.
	- Updated existing tests to enhance maintainability and address threading model changes.

### CODE REVIEW

**Comment by maciejmajek** [comment]
@coderabbit ai full review

**Comment by maciejmajek** [comment]
@coderabbitai full review

File: src/rai/rai/communication/ros2/connectors.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+from rai.communication.ros2.api import ROS2ActionAPI, ROS2ServiceAPI, ROS2TopicAPI
+
+
+class ROS2ARIMessage(ARIMessage):
+    def __init__(self, payload: Any, metadata: Optional[Dict[str, Any]] = None):
+        super().__init__(payload, metadata)
+
+
+class ROS2ARIConnector(ARIConnector[ROS2ARIMessage]):
+    def __init__(self, node_name: str = "rai_ros2_ari_connector"):
```
Comment (rachwalk) [inline src/rai/rai/communication/ros2/connectors.py]: The question is should this name be default, as it can create conflicts in the network. Perhaps something like `f"rai_ros2_ari_connector_{uuid4()[:6]}"` could be more sensible if we want this default argument to be functional?

File: src/rai/rai/communication/ros2/connectors.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        msg_type = message.metadata.get("msg_type", None)
+
+        # TODO: allow msg_type to be None, add auto topic type detection
+        if msg_type is None:
+            raise ValueError("msg_type is required")
+
+        self._topic_api.publish(
+            topic=target,
+            msg_content=message.payload,
+            msg_type=message.metadata["msg_type"],
```
Comment (rachwalk) [inline src/rai/rai/communication/ros2/connectors.py]:
```suggestion
            msg_type=msg_type
```

File: src/rai/rai/communication/ros2/connectors.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    def terminate_action(self, action_handle: str):
+        self._actions_api.terminate_goal(action_handle)
+
+    def shutdown(self):
+        self._cleanup()
+
+    def _cleanup(self):
+        self._executor.shutdown()
+        self._thread.join()
+        self._node.destroy_node()
```
Comment (rachwalk) [inline src/rai/rai/communication/ros2/connectors.py]: What's the benefit of having a separate _cleanup() function?

File: tests/communication/ros2/test_connectors.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        handle = connector.start_action(
+            action_data=message,
+            target=action_name,
+            on_feedback=lambda feedback: feedbacks.append(feedback),
+        )
+        assert handle is not None
+        feedbacks_before = feedbacks
+        connector.terminate_action(handle)
+        feedbacks_after = feedbacks
+        assert len(feedbacks_after) == len(feedbacks_before)
```
Comment (rachwalk) [inline tests/communication/ros2/test_connectors.py]: Why is it anticipated that there will be no feedback published? It seems counterintuitive that a successfull action call creates no feedback (the main benefit of an action over a service).

**Review by rachwalk — COMMENTED** [review]
I  have a few questions regarding design choices, would appreciate if you address them

File: src/rai/rai/communication/ros2/connectors.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+            action_type=action_data.metadata["msg_type"],
+            goal=action_data.payload,
+            timeout_sec=timeout_sec,
+            feedback_callback=on_feedback,
+            done_callback=on_done,
+        )
+        if not accepted:
+            raise RuntimeError("Action goal was not accepted")
+        return handle
+
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/connectors.py]: Unfortunately, the action_data has to be optional in order to keep the types correct. Here is the pylance error
```
Method "start_action" overrides class "BaseConnector" in an incompatible manner
  Parameter 2 type mismatch: base parameter is type "ROS2ARIMessage | None", override parameter is type "ROS2ARIMessage"
    Type "ROS2ARIMessage | None" is incompatible with type "ROS2ARIMessage"
      "None" is incompatible with "ROS2ARIMessage"Pylance[reportIncompatibleMethodOverride](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportIncompatibleMethodOverride)
 ```
 Are there any alternatives?

File: src/rai/rai/communication/ros2/api.py [hunk 6]
```diff
@@ -324,18 +360,52 @@ def __init__(self, node: rclpy.node.Node) -> None:
         self.node = node
         self._logger = node.get_logger()
         self.actions: Dict[str, ROS2ActionData] = {}
+        self._callback_executor = ThreadPoolExecutor(max_workers=10)
 
     def _generate_handle(self):
         return str(uuid.uuid4())
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/api.py:366]: Done

File: src/rai/rai/communication/ros2/api.py [hunk 7]
```diff
@@ -403,3 +479,8 @@ def get_result(self, handle: str) -> Any:
         if self.actions[handle]["result_future"] is None:
             raise ValueError(f"No result available for goal {handle}")
         return self.actions[handle]["result_future"].result()
+
+    def __del__(self) -> None:
+        """Cleanup thread pool when object is destroyed."""
+        if hasattr(self, "_callback_executor"):
+            self._callback_executor.shutdown(wait=False)
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/api.py]: Done

File: tests/communication/ros2/test_connectors.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    message_receiver = MessageReceiver(topic_name)
+    executors, threads = multi_threaded_spinner([message_receiver])
+    connector = ROS2ARIConnector()
+    try:
+        message = ROS2ARIMessage(
+            payload={"data": "Hello, world!"},
+            metadata={"msg_type": "std_msgs/msg/String"},
+        )
+        connector.send_message(message, target=topic_name)
+        time.sleep(1)  # wait for the message to be received
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply tests/communication/ros2/test_connectors.py:50]: This will be done in the future. The tests are actually kind of edge case for the discovery mechanism, as most of the time the topics are static and well discovered before running any method of the api

File: tests/communication/ros2/test_connectors.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+from .helpers import (
+    MessagePublisher,
+    MessageReceiver,
+    ServiceServer,
+    multi_threaded_spinner,
+    ros_setup,
+    shutdown_executors_and_threads,
+)
+
+ros_setup  # type: ignore prevents pytest from complaining about unused fixture
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply tests/communication/ros2/test_connectors.py]: Done

File: tests/communication/ros2/test_api.py [hunk 10]
```diff
@@ -14,121 +14,26 @@
...
-    for thread in threads:
-        thread.join(timeout=2.0)
-
-
-@pytest.fixture(scope="function")
-def ros_setup() -> Generator[None, None, None]:
-    rclpy.init()
-    yield
-    rclpy.shutdown()
+ros_setup  # type: ignore prevents pytest from complaining about unused fixture
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply tests/communication/ros2/test_api.py]: Done

File: src/rai/rai/communication/base_connector.py [hunk 11]
```diff
@@ -13,15 +13,26 @@
...
+        payload: Any,
+        metadata: Optional[Dict[str, Any]] = None,
+        *args: Any,
+        **kwargs: Any,
+    ):
         self.payload = payload
+        if metadata is None:
+            self.metadata = {}
+        else:
+            self.metadata = metadata
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/communication/base_connector.py:35]: Please use numpydoc formatting for suggestion

File: tests/communication/ros2/test_connectors.py (hunk 4 above)
Comment (maciejmajek) [reply tests/communication/ros2/test_connectors.py]: After terminating the action you are not supposed to receive any kind of feedback message.

File: src/rai/rai/communication/ros2/connectors.py (hunk 1 above)
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/connectors.py]: Good catch

File: tests/communication/ros2/test_connectors.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        connector.shutdown()
+        shutdown_executors_and_threads(executors, threads)
+
+
+@pytest.mark.xfail(
+    reason="Test expected to fail: ROS2 node discovery is asynchronous and the current implementation "
+    "doesn't wait for topic discovery. "
+    "TODO: Implement a proper discovery mechanism with timeout "
+    "to ensure reliable topic communication."
+)
```
[the remark this answers is not in the document]
Comment (maciejmajek) [reply tests/communication/ros2/test_connectors.py:62]: Will be done in the future.

File: src/rai/rai/communication/ros2/connectors.py (hunk 3 above)
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/connectors.py]: None, refactored

File: src/rai/rai/communication/ros2/connectors.py (hunk 1 above)
Comment (maciejmajek) [reply src/rai/rai/communication/ros2/connectors.py]: Done

File: src/rai/rai/communication/base_connector.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (maciejmajek) [reply src/rai/rai/communication/base_connector.py:35]: Thanks, will be done once we establish docstring format

**Review by rachwalk — APPROVED** [review]
LGTM
